core/sr-primitives/src/generic/unchecked_mortal_compact_tipped_extrinsic.rs

```diff
@@ -0,0 +1,467 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Generic implementation of an unchecked (pre-verification) extrinsic with tip.
+
+#[cfg(feature = "std")]
+use std::fmt;
+
+use rstd::prelude::*;
+use runtime_io::blake2_256;
+use crate::codec::{Decode, Encode, Input, Compact};
+use crate::traits::{self, Member, SimpleArithmetic, MaybeDisplay, CurrentHeight, BlockNumberToHash,
+	Lookup, Checkable, Extrinsic, SaturatedConversion};
+use super::{CheckedExtrinsic, Era, Tip};
+
+const TRANSACTION_VERSION: u8 = 2;
+
+/// A extrinsic right from the external world. This is unchecked and so
+/// can contain a signature.
+///
+/// This type transaction:
+///   - _Must_ always have a valid `Tip` encoded, and included in the signature payload if it is
+///   signed.
+///   - _Must_ not provide any `Tip` if it is unsigned.
+#[derive(PartialEq, Eq, Clone)]
+pub struct UncheckedMortalCompactTippedExtrinsic<Address, Index, Call, Signature, Balance> {
+	/// The signature, address, number of extrinsics have come before from
+	/// the same signer and an era describing the longevity of this transaction,
+	/// if this is a signed extrinsic.
+	pub signature: Option<(Address, Signature, Compact<Index>, Era)>,
+	/// The function that should be called.
+	pub function: Call,
+	/// The tip for this transaction
+	pub tip: Option<Tip<Balance>>,
+}
+
+impl<Address, Index, Call, Signature, Balance>
+	UncheckedMortalCompactTippedExtrinsic<Address, Index, Call, Signature, Balance>
+{
+	/// New instance of a signed extrinsic aka "transaction".
+	pub fn new_signed(
+		index: Index,
+		function: Call,
+		signed: Address,
+		signature: Signature,
+		era: Era,
+		tip: Option<Tip<Balance>>
```

Comment:
In my head it makes sense; a signed can say _my tip is none_ (which is practically the same as saying my tip is `Sender(0)`) but it is still fine as long as __it is provided in the byte stream and it can be decoded__ (afaik `None` is `0` or `0 0` when encoded?).

Reply:
ignore the above ^^  you're right.